File: ts/src/resources/activepieces.ts
import { HttpTransport } from "../http.js"

export class ActivePiecesResource {
  /**
   * @param base - ActivePieces base URL (gateway/activepieces)
   */
  constructor(private readonly http: HttpTransport, private readonly base: string) {}  

  /**
   * Generic GET request to ActivePieces backend
   */
  async get(path: string, params?: Record<string, string>): Promise<any> {
    const url = new URL(`${this.base}${path.startsWith('/') ? path : `/${path}`}`)    
    if (params) {
      Object.entries(params).forEach(([k, v]) => url.searchParams.set(k, v))
    }
    return this.http.getFetch()(url, { method: "GET" }).then(r => r.json())
  }

  /**
   * Generic POST request to ActivePieces backend
   */
  async post(path: string, body: any): Promise<any> {
    const url = `${this.base}${path.startsWith('/') ? path : `/${path}`}`
    return this.http.getFetch()(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  /**
   * Generic PUT request to ActivePieces backend
   */
  async put(path: string, body: any): Promise<any> {
    const url = `${this.base}${path.startsWith('/') ? path : `/${path}`}`
    return this.http.getFetch()(url, {
      method: "PUT",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(body),
    }).then(r => r.json())
  }

  /**
   * Generic DELETE request to ActivePieces backend
   */
  async delete(path: string): Promise<any> {
    const url = `${this.base}${path.startsWith('/') ? path : `/${path}`}`
    return this.http.getFetch()(url, { method: "DELETE" }).then(r => r.json())
  }
}

File: py/src/imbrace/resources/activepieces.py
from typing import Any, Dict, Optional
from ..http import HttpTransport, AsyncHttpTransport

class ActivePiecesResource:
    """ActivePieces Backend Proxy — Sync."""

    def __init__(self, http: HttpTransport, base: str):
        self._http = http
        self._base = base.rstrip('/')

    def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        return self._http.request("GET", full_path, params=params or {}).json()

    def post(self, path: str, body: Any) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        return self._http.request("POST", full_path, json=body).json()

    def put(self, path: str, body: Any) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        return self._http.request("PUT", full_path, json=body).json()

    def delete(self, path: str) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        return self._http.request("DELETE", full_path).json()


class AsyncActivePiecesResource:
    """ActivePieces Backend Proxy — Async."""

    def __init__(self, http: AsyncHttpTransport, base: str):
        self._http = http
        self._base = base.rstrip('/')

    async def get(self, path: str, params: Optional[Dict[str, Any]] = None) -> Any:    
        full_path = f"{self._base}/{path.lstrip('/')}"
        res = await self._http.request("GET", full_path, params=params or {})
        return res.json()

    async def post(self, path: str, body: Any) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        res = await self._http.request("POST", full_path, json=body)
        return res.json()

    async def put(self, path: str, body: Any) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        res = await self._http.request("PUT", full_path, json=body)
        return res.json()

    async def delete(self, path: str) -> Any:
        full_path = f"{self._base}/{path.lstrip('/')}"
        res = await self._http.request("DELETE", full_path)
        return res.json()
